File: server/dukan/domain/inventory.py
"""Sample domain slice. Mirrors docs/domain/inventory.md and the Dart
implementation in packages/dukan_core/lib/domain/inventory.dart — the SAME test
table is applied in both languages (see tests/unit/test_inventory.py).

Pure: imports nothing but the shared error contract. No framework, no I/O, no
datetime.now() (a time value is passed in).
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum

from dukan.shared.errors import ConflictError, ValidationError


class StockReason(StrEnum):
    SALE = "sale"
    PURCHASE = "purchase"
    ADJUSTMENT = "adjustment"
    TRANSFER_IN = "transfer_in"
    TRANSFER_OUT = "transfer_out"
    COUNT = "count"
    RETURNED = "returned"


@dataclass(frozen=True, slots=True)
class StockMovement:
    """Append-only, immutable stock ledger entry."""

    id: str
    product_id: str
    branch_id: str
    qty_delta: int  # signed: negative for a sale
    reason: StockReason
    occurred_at: datetime


def on_hand(movements: Iterable[StockMovement]) -> int:
    """On-hand = sum of movement deltas."""
    return sum(m.qty_delta for m in movements)


def sell_from_stock(
    *,
    id: str,
    product_id: str,
    branch_id: str,
    qty: int,
    available: int,
    at: datetime,
) -> StockMovement:
    """Produce the sale movement, enforcing the single-device/online oversell rule.

    Raises ValidationError(STOCK_INVALID_QTY) or ConflictError(STOCK_INSUFFICIENT).
    """
    if qty <= 0:
        raise ValidationError("STOCK_INVALID_QTY", qty=qty)
    if qty > available:
        raise ConflictError("STOCK_INSUFFICIENT", requested=qty, available=available)
    return StockMovement(
        id=id,
        product_id=product_id,
        branch_id=branch_id,
        qty_delta=-qty,
        reason=StockReason.SALE,
        occurred_at=at,
    )


def adjust_stock(
    *,
    id: str,
    product_id: str,
    branch_id: str,
    qty_delta: int,
    at: datetime,
) -> StockMovement:
    """Manual stock adjustment; qty_delta is signed and must be non-zero.

    Raises ValidationError(STOCK_INVALID_QTY) when qty_delta == 0.
    """
    if qty_delta == 0:
        raise ValidationError("STOCK_INVALID_QTY", qty=qty_delta)
    return StockMovement(
        id=id,
        product_id=product_id,
        branch_id=branch_id,
        qty_delta=qty_delta,
        reason=StockReason.ADJUSTMENT,
        occurred_at=at,
    )

File: packages/dukan_core/lib/domain/inventory.dart
import '../shared/errors.dart';

/// Sample domain slice anchoring the layer and mirroring
/// `docs/domain/inventory.md`. Pure, framework-free, testable with zero
/// fixtures. Real catalog/sales aggregates land in Phases 2–3 alongside it.

enum StockReason { sale, purchase, adjustment, transferIn, transferOut, count, returned }

/// An append-only, immutable stock ledger entry. On-hand is **derived** by
/// summing these — never stored as a mutable running total.
final class StockMovement {
  const StockMovement({
    required this.id,
    required this.productId,
    required this.branchId,
    required this.qtyDelta,
    required this.reason,
    required this.occurredAt,
  });

  final String id;
  final String productId;
  final String branchId;
  final int qtyDelta; // signed: negative for a sale
  final StockReason reason;
  final DateTime occurredAt;
}

/// On-hand for a product in a branch = Σ of its movement deltas.
int onHand(Iterable<StockMovement> movements) =>
    movements.fold(0, (sum, m) => sum + m.qtyDelta);

/// Produce the sale movement for selling [qty] units, enforcing the
/// single-device / online oversell rule. Offline partition handling (detect,
/// not prevent) is a sync concern; see `docs/sync-protocol.md`.
///
/// Raises:
/// - [ValidationError] `STOCK_INVALID_QTY` if qty <= 0
/// - [ConflictError] `STOCK_INSUFFICIENT` if qty > available
StockMovement sellFromStock({
  required String id,
  required String productId,
  required String branchId,
  required int qty,
  required int available,
  required DateTime at,
}) {
  if (qty <= 0) {
    throw ValidationError('STOCK_INVALID_QTY', {'qty': qty});
  }
  if (qty > available) {
    throw ConflictError(
        'STOCK_INSUFFICIENT', {'requested': qty, 'available': available});
  }
  return StockMovement(
    id: id,
    productId: productId,
    branchId: branchId,
    qtyDelta: -qty,
    reason: StockReason.sale,
    occurredAt: at,
  );
}
